File: programmers/01.hash/05.베스트앨범/js/01.js
/*
많은 장르 먼저 
    장르 내  많이 재생
        고유번호 낮은 노래


*/

/*
map = {
    장르 : [[재생횟수, 번호]]
}
*/

function solution(genres, plays) {
    let answer = [];
    const mapping = genres.reduce((map, now, i) => {
        const arr = map[now] ? map[now] : [0];
        arr[0] += plays[i];
        arr.push([plays[i], i]);
        map[now] = arr;
        return map;
    }, {});

    const arr = Object.keys(mapping).map((genres) => {
        mapping[genres].sort((a, b) => b[0] - a[0]);
        return [genres, mapping[genres][0]];
    });
    arr.sort((a, b) => b[1] - a[1]);

    arr.forEach((genres) =>
        mapping[genres[0]].slice(1, 3).map((sing) => answer.push(sing[1]))
    );
    return answer;
}

// test
console.log(
    solution(["classic", "pop", "classic", "classic", "pop"], [500, 600, 150, 800, 2500])
);

File: programmers/01.hash/05.베스트앨범/python/01.py
# 왜틀리는거지??
def solution(genres, plays):
    answer = []

    mapping = {}
    for i, (genre, play) in enumerate(zip(genres, plays)):
        arr = mapping[genre] if genre in mapping else [0]
        arr[0] += play
        arr += [[play, i]]
        mapping[genre] = arr

    for (genre, list) in sorted(mapping.items(), key= lambda x : x[0], reverse=True):
        for (play, idx) in sorted(list[1:], key= lambda x : x[0], reverse=True)[:2]:
            answer.append(idx)
    
    return answer



#test
print(solution(["classic", "pop", "classic", "classic", "pop", "test"],[500, 600, 150, 800, 2500, 8000]))